ddtrace/tracer/sampler.go

```diff
@@ -158,49 +158,115 @@ func (ps *prioritySampler) apply(spn *span) {
+// singleSpanRulesSampler samples individual spans based on a separate user-defined set of rules and
+// cannot impact the trace sampling decision.
 type rulesSampler struct {
-	rules      []SamplingRule // the rules to match spans with
-	globalRate float64        // a rate to apply when no rules match a span
-	limiter    *rateLimiter   // used to limit the volume of spans sampled
+	traceRulesSampler      *traceRulesSampler
+	singleSpanRulesSampler *singleSpanRulesSampler
 }
 
 // newRulesSampler configures a *rulesSampler instance using the given set of rules.
+// Rules are split between trace and single span sampling rules according to their type.
+// Such rules are user-defined through environment variable or WithSamplingRules option.
 // Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
 func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	var spanRules, traceRules []SamplingRule
+	for _, rule := range rules {
+		if rule.Type == SamplingRuleSingleSpan {
+			rule.limiter = newSingleSpanRateLimiter(rule.MaxPerSecond)
+			spanRules = append(spanRules, rule)
+		} else {
+			traceRules = append(traceRules, rule)
+		}
+	}
 	return &rulesSampler{
-		rules:      rules,
-		globalRate: globalSampleRate(),
-		limiter:    newRateLimiter(),
+		traceRulesSampler:      newTraceRulesSampler(traceRules),
+		singleSpanRulesSampler: newSingleSpanRulesSampler(spanRules),
 	}
 }
 
-// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES
-// environment variable.
+const (
+	// SamplingRuleTrace specifies that if a span matches a sampling rule of this type,
+	// an entire trace might be kept based on the given span and sent to the agent.
+	SamplingRuleTrace = iota
+	// SamplingRuleSingleSpan specifies that if a span matches a sampling rule,
+	// an individual span might be kept and sent to the agent.
+	SamplingRuleSingleSpan
+)
+
+// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES,
+// DD_SPAN_SAMPLING_RULES and DD_SPAN_SAMPLING_RULES_FILE environment variables.
 func samplingRulesFromEnv() ([]SamplingRule, error) {
+	var errs []string
+	var rules []SamplingRule
 	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		traceRules, err := processSamplingRules([]byte(rulesFromEnv), SamplingRuleTrace)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		rules = append(rules, traceRules...)
+	}
+	spanRules, err := spanSamplingRulesFromEnv()
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	rules = append(rules, spanRules...)
+	if len(errs) != 0 {
+		return rules, fmt.Errorf("%s", strings.Join(errs, "\n\t"))
+	}
+	return rules, nil
+}
+
+// spanSamplingRulesFromEnv parses sampling rules from the DD_SPAN_SAMPLING_RULES and
+// DD_SPAN_SAMPLING_RULES_FILE environment variables.
+func spanSamplingRulesFromEnv() ([]SamplingRule, error) {
```

Comment:
I would absorb this function into the one above. It's a bit too much. Do we need both? They aren't really used anywhere and they don't necessarily make figuring this out easier.

Reply:
I wanted too but didn't like how it looked. I guess I can try again 🙂